### PR TITLE
Keyboard dialog and shortcuts

### PR DESCRIPTION
**Pull request by iafan**
Implement Mac-compatible shortcuts and keyboard help dialog

### CODE REVIEW

File: pootle/static/js/shared/components/KeyboardHelpDialog.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    const enterKey = <kbd>Enter</kbd>;
+    /* const escKey = <kbd>Esc</kbd>; */
+    const spaceKey = <kbd>Space</kbd>;
+    const or = <div className="or">or</div>;
+
+    return (
+      <div>
+        <div className="column">
+
+          <div className="section">Editor</div>
```
Comment (julen) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: i18n is missing here (and in many other places).

File: pootle/static/js/common.js [hunk 2]
```diff
@@ -203,6 +205,16 @@ PTL.common = {
       });
     });
 
+
+    /* Bind hotkeys */
+    const hotkeys = mousetrap(document.body);
+
+    hotkeys.bind('f1', (e) => {
```
Comment (julen) [inline pootle/static/js/common.js]: I think `F1` is a Windows convention. In web apps, I have always found <kbd>?</kbd> as the trigger. I am aware that <kbd>F1</kbd> also triggers while textareas/inputs have focus, although I find it a bit counter-intuitive as a non-Windows user. How about mapping <kbd>?</kbd> in addition to <kbd>F1</kbd>?

File: pootle/static/js/shared/components/KeyboardHelpDialog.js [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+export function showKeyboardHelpDialog() {
+  if (keyboardHelpOpen) {
+    return;
+  }
+
+  keyboardHelpOpen = true;
+  showModal({
+    title:
+      <span>
+        Keyboard Help (press <kbd>F1</kbd> to open;
```
Comment (julen) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: Isn't the text in parentheses a bit redundant? One can currently only open the dialog via the keyboard only, and `Esc`, as being universally understood, is a shortcut that doesn't deserve explicit mention IMHO.

**Comment by julen** [comment]
I would suggest to also update the existing docs to don't lose track of existing shortcuts. An entry in the changelog is well-deserved too.

File: pootle/static/js/shared/components/KeyboardHelpDialog.js (hunk 3 above)
Comment (iafan) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: I definitely plan to have a way to open the dialog from somewhere else — I don't want it to be only accidentally discovered by pressing `F1`. This is why I believe it is important to have this text. Also, this dialog might be the first time the user sees any mention of keyboard shortcuts, so I think it is the right time to educate them that they can use `Esc` to close this dialog (and give a hint that this also works in other popups). I wouldn't rely on "Esc being universally understood", especially when it comes to web apps.

File: pootle/static/js/common.js (hunk 2 above)
Comment (iafan) [inline pootle/static/js/common.js]: One should be able to invoke keyboard help dialog while being inside the input box (which is what users will have by default in translation mode). So bare <kbd>?</kbd> is not an option. We can try <kbd>⌘</kbd>+<kbd>?</kbd> or <kbd>Control</kbd>+<kbd>?</kbd> or whatever works on Mac and doesn't clash with input fields.

File: pootle/static/js/shared/components/KeyboardHelpDialog.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                {modKey}<kbd>&darr;</kbd>
+                {or}
+                {ctrlKey}<kbd>.</kbd>
+              </td>
+            </tr>
+            <tr>
+              <td>Submit/suggest + go to next unit</td>
+              <td>{modKey}{enterKey}</td>
+            </tr>
+            {/*
```
Comment (iafan) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: OK. Thanks for having it removed, but now I know I can always go back to the first revision in Git history to get what I need. :)

File: pootle/static/js/shared/components/KeyboardHelpDialog.js (hunk 3 above)
Comment (julen) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: How about adding these under the _General_ section instead? I don't think these belong there, but I am happy to change my opinion if you show me examples of popular web apps that display such shortcuts in the title.

File: pootle/static/js/common.js (hunk 2 above)
Comment (julen) [inline pootle/static/js/common.js]: <kbd>⌘</kbd>+<kbd>?</kbd> is reserved for the OS' help menu and I wouldn't resource to using <kbd>Ctrl</kbd> at all as it's not in line with the platform.

To be honest, I don't think there's anything wrong with mapping it to <kbd>?</kbd> additionaly, even if it cannot be triggered from a focused editor.

File: pootle/static/js/shared/components/KeyboardHelpDialog.js (hunk 3 above)
Comment (iafan) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: See e.g. Inoreader. They don't have a _title_ per se, as they render this in a footer, but still, the approach is the same — to visually separate how the help dialog can be invoked, from the rest of the shortcuts:

Image

**Comment by iafan** [comment]
I updated both changelog (32348b6717e11d4d9b2b6bd8a63f7e8f71fa2055) and keyboard documentation (fc7492993cf794ce1b463ea654a2bb1633edd753) accordingly.

File: pootle/static/js/common.js (hunk 2 above)
Comment (iafan) [inline pootle/static/js/common.js]: Nothing wrong with that being an alternative shortcut. I added it in 237982bda9d88ee0c88500f7c5a2ae396e29e549

File: pootle/static/js/shared/components/KeyboardHelpDialog.js (hunk 3 above)
Comment (julen) [inline pootle/static/js/shared/components/KeyboardHelpDialog.js]: As you see _Close dialog_ is a generic one (like in Zing), and as such it is listed under the global section. This is what other apps such as Trello do too. Therefore I would move that down to the general section.

pantaila-argazkia 2016-10-26 10 50 25

**Comment by iafan** [comment]
> Therefore I would move that down to the general section.

Implemented in 4865b63a6cc7e6e455be2c14f76a88d14ecef941

**Pull request merged**